data_safe_haven/pulumi/components/sre_research_desktop.py

```diff
@@ -40,21 +44,21 @@ def __init__(
         self.location = location
         self.security_group_name = security_group_name
         self.virtual_network_name = Output.from_input(virtual_network).apply(
-            lambda v: v.name
+            get_name_from_vnet
         )
         self.subnet_research_desktops_name = Output.from_input(
             subnet_research_desktops
-        ).apply(lambda s: s.name)
+        ).apply(get_name_from_subnet)
         self.virtual_network_resource_group_name = Output.from_input(
             virtual_network_resource_group
-        ).apply(lambda rg: rg.name)
+        ).apply(get_name_from_rg)
         self.vm_ip_addresses = Output.all(subnet_research_desktops, vm_details).apply(
             lambda args: self.get_ip_addresses(
-                cast(network.GetSubnetResult, args[0]), len(cast(list, args[1]))
+                cast(network.GetSubnetResult, args[0]), len(cast(List[str], args[1]))
```

Comment:
Is this one avoidable?

Reply:
It's difficult. The only type that `mypy` can infer for `args` here is `List[network.GetSubnetResult | List[str]]`. This means that we can't pass `args[0]` and `args[1]` to `get_ip_addresses` since this expects one `network.GetSubnetResult` and one `List[str]`.

The only rewrite I can think of is to change `get_ip_addresses` to accept `List[network.GetSubnetResult | List[str]]` directly and do the type checking inside that function. It's a bit messy though and means that the conversion will be done in a different place than the original obfuscation.

Maybe you can think of another option?

Reply:
I suppose `Output.all()` is required to combine the two outputs into a single new output? Otherwise Pulumi can't ensure both promises have values?

In that case, `get_ip_addresses` could take arguments of the union type, but that seems bad because neither input is really a union. Feels like a case where it is appropriate to use Any and rely on duck typing.

I have opened a PR on your fork with a suggestion https://github.com/jemrobinson/data-safe-haven/pull/2